### PR TITLE
feat: add STFT->MEL conversion

### PR DESCRIPTION
**Pull request by polvalente**

### CODE REVIEW

File: lib/nx_signal.ex [hunk 1]
```diff
@@ -165,44 +201,77 @@ defmodule NxSignal do
...
+      case opts[:stride] do
+        stride when Elixir.Kernel.is_list(stride) ->
+          stride
+
+        stride
+        when Elixir.Kernel.and(
+               Elixir.Kernel.is_integer(stride),
+               Elixir.Kernel.>=(stride, 1)
+             ) ->
+          [stride]
```
Comment (josevalim) [inline lib/nx_signal.ex]: Why are you using `Elixir.Kernel` here? :)

File: lib/nx_signal.ex [hunk 2]
```diff
@@ -165,44 +201,77 @@ defmodule NxSignal do
...
+
+        stride ->
+          raise ArgumentError,
+                "expected an integer >= 1 or a list of integers, got: #{inspect(stride)}"
+      end
+
+    dilations = List.duplicate(1, Nx.rank(shape))
+
+    {pooled_shape, padding_config} =
+      Nx.Shape.pool(shape, window_dimensions, strides, padding, dilations)
```
Comment (josevalim) [inline lib/nx_signal.ex]: Private API :(

File: lib/nx_signal.ex [hunk 3]
```diff
@@ -297,4 +467,155 @@ defmodule NxSignal do
...
+      )
+
+    mel_spec = Nx.reshape(mel_spec, Nx.shape(mel_spec), names: [:frames, :mel])
+
+    log_spec = Nx.log(Nx.clip(mel_spec, 1.0e-10, :infinity)) / Nx.log(10)
+    log_spec = Nx.max(log_spec, Nx.reduce_max(log_spec) - 8)
+    (log_spec + 4) / 4
+  end
+
+  defn linspace(min, max, opts \\ []) do
```
Comment (josevalim) [inline lib/nx_signal.ex]: Public or private?

File: lib/nx_signal.ex [hunk 4]
```diff
@@ -297,4 +467,155 @@ defmodule NxSignal do
...
+        [
+          [0.0, 8.129207417368889e-4, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+          [0.0, 9.972017724066973e-4, 2.1870265481993556e-4, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+          [0.0, 0.0, 9.510896052233875e-4, 4.1505080298520625e-4, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+          [0.0, 0.0, 0.0, 4.035892488900572e-4, 5.276654846966267e-4, 2.5741232093423605e-4, 0.0, 0.0, 0.0, 0.0],
+          [0.0, 0.0, 0.0, 0.0, 7.329042273340747e-5, 2.3422071535605937e-4, 3.8295125705190003e-4, 2.8712046332657337e-4, 1.9128969870507717e-4, 9.545891225570813e-5]
+        ]
+      >
+  """
+  defn mel_filters(opts \\ []) do
```
Comment (josevalim) [inline lib/nx_signal.ex]: You should make this function a `deftransform` and all numeric values that are given as options should be passed as regular arguments to the underlying `defnp mel_filters_impl` :)

File: lib/nx_signal.ex [hunk 5]
```diff
@@ -60,26 +62,58 @@ defmodule NxSignal do
...
+
+  ## Examples
+
+      iex> NxSignal.fft_frequencies(nfft: 10, fs: 1.6e4)
+      #Nx.Tensor<
+        f32[frequencies: 10]
+        [0.0, 1.6e3, 3.2e3, 4.8e3, 6.4e3, 8.0e3, 9.6e3, 1.12e4, 1.28e4, 1.44e4]
+      >
+  """
+  defn fft_frequencies(opts \\ []) do
```
Comment (josevalim) [inline lib/nx_signal.ex]: We should make `fs` a regular argument here. :)

File: lib/nx_signal.ex [hunk 6]
```diff
@@ -39,19 +40,20 @@ defmodule NxSignal do
...
+        f32[frequencies: 2]
         [0.0, 200.0]
       >
   """
   defn stft(data, window, opts \\ []) do
     {frame_size} = Nx.shape(window)
 
-    {overlap_size, fs} =
+    {overlap_size, fs, padding} =
       transform({frame_size, opts}, fn {frame_size, opts} ->
```
Comment (josevalim) [inline lib/nx_signal.ex]: Consider implement `stft` as `deftransform` and passing `fs` as an argument to `defn`, so it is properly converted to a tensor!

File: lib/nx_signal.ex (hunk 1 above)
Comment (polvalente) [reply lib/nx_signal.ex]: Probably this was a `transform` before with inline `fn`

Comment (polvalente) [reply lib/nx_signal.ex]: No need for this now

File: lib/nx_signal.ex [hunk 7]
```diff
@@ -224,6 +301,107 @@ defmodule NxSignal do
     output
   end
 
+  defnp as_windowed_reflect_padding(tensor, opts \\ []) do
```
Comment (seanmor5) [inline lib/nx_signal.ex:341]: Reflect padding might be a good addition upstream to Nx

**Review by seanmor5 — APPROVED** [review]

File: lib/nx_signal.ex (hunk 3 above)
Comment (seanmor5) [reply lib/nx_signal.ex]: linspace is also used a lot in Bumblebee, it may be worth considering for Nx upstream as well

Comment (polvalente) [reply lib/nx_signal.ex]: +1 for that. It's a really common function, albeit a short one

File: lib/nx_signal.ex (hunk 7 above)
Comment (polvalente) [reply lib/nx_signal.ex:341]: The main difficulty with that is that currently padding will create values that are repeated (through {left, right, inner} padding config tuples).

I can't see how we'd frame this in terms of that API

Comment (polvalente) [reply lib/nx_signal.ex:341]: I mean the API that talks to backends, not the public-facing one

File: lib/nx_signal.ex [hunk 8]
```diff
@@ -297,4 +481,151 @@ defmodule NxSignal do
...
+  ## Arguments
+    * `z` - STFT spectrum
+    * `fs` - Sampling frequency in Hz
+
+  ## Options
+
+    * `:nfft` - Number of FFT bins
+    * `:nmels` - Number of target MEL bins. Defaults to 128
+    * `:type` - Target output type. Defaults to `{:f, 32}`
+
```
Comment (jonatanklosko) [inline lib/nx_signal.ex:610]: What about more explicit names, like:

* `fs` -> `sampling_frequency`/`sampling_rate`
* `nfft` -> `num_fft_bins`
* `nmels` -> `num_mel_bins`

Comment (polvalente) [reply lib/nx_signal.ex:610]: fs and nfft specifically are usual names for these params, that's why I chose them, but nmels not so much

https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.stft.html
https://pytorch.org/docs/stable/generated/torch.stft.html
https://www.mathworks.com/help/signal/ref/stft.html

WDYT about:
nfft  -> fft_length
fs -> sampling_frequency
nmels -> mel_bins

Comment (polvalente) [reply lib/nx_signal.ex:610]: nfft/fs would need to change for more functions I think

Comment (jonatanklosko) [reply lib/nx_signal.ex:610]: FFT is itself an acronym, so adding n on top of that may easily confuse someone without much prior experience, so a more explicit name is generally helpful.

The names above sound good to me!

Comment (polvalente) [reply lib/nx_signal.ex:610]: Cool, I'll change these!

Comment (polvalente) [reply lib/nx_signal.ex:610]: for consistency, I'm also changing overlap_size -> overlap_length given that both it and nfft operate on the same headspace of "samples"

**Comment by polvalente** [comment]
Because this PR has already taken too many extraneous changes, I'll leave the refactor for the reflection padding to use Nx.reflect for a follow-up PR

**Review by josevalim — APPROVED** [review]

**Pull request merged**